backend/buildeR.BLL/Services/TeamMemberService.cs

```diff
@@ -1,41 +1,166 @@
+            INotificationsService notificationsService) : base(context, mapper) 
+        {
+            _context = context;
+            _userService = userService;
+            _emailBuilder = emailBuilder;
+            _emailService = emailService;
+            _notificationsService = notificationsService;
+        }
         public async Task<TeamMemberDTO> Create(NewTeamMemberDTO teamMember)
         {
             if (teamMember == null)
             {
                 throw new ArgumentNullException();
             }
             teamMember.JoinedDate = DateTime.Now;
-            return await base.AddAsync(teamMember);
+
+            var user = await _context.Users.Include(u => u.NotificationSetting).FirstOrDefaultAsync(u => u.Id == teamMember.UserId);
+            //if (user.NotificationSetting.EnableEmail)
+            //{
+            //    var emailModel = _emailBuilder.GetInviteGroupLetter(user.Email, user.FirstName);
+            //    await _emailService.SendEmailAsync(new List<string> { emailModel.Email }, emailModel.Subject, emailModel.Title, emailModel.Body);
+            //}
+
+            var newMember = await base.AddAsync(teamMember);
+
+            var inviterUsername = (await _context.Users.FirstOrDefaultAsync(u => u.Id == teamMember.InitiatorId)).Username;
+            var groupName = (await _context.Groups.FirstOrDefaultAsync(g => g.Id == teamMember.GroupId)).Name;
+
+            await _notificationsService.Create(new NewNotificationDTO
+            {
+                Message = $"{inviterUsername} invites you to join group {groupName}",
+                Date = DateTime.Now,
+                Type = NotificationType.Group,
+                UserId = teamMember.UserId,
+                ItemId = -1 // after click on notification user will see Groups page
+            });
+
+            return newMember;
         }
 
-        public Task Delete(int id)
+        public async Task Delete(int id)
         {
-            throw new NotImplementedException();
+            var member = await base.GetAsync(id);
+            if (member == null)
+            {
+                throw new NotFoundException(nameof(TeamMember), id);
+            }
+
+            await base.RemoveAsync(id);
+        }
+
+        public async Task DeleteWithNotification(RemoveTeamMemberDTO teamMember)
+        {
+            var member = await base.GetAsync(teamMember.Id);
+            int memberUserId = member.UserId;
+            bool memberIsAccepted = member.IsAccepted;
+            if (member == null)
+            {
+                throw new NotFoundException(nameof(TeamMember), teamMember.Id);
+            }
+            await base.RemoveAsync(teamMember.Id);
+
+            var groupName = (await _context.Groups.FirstOrDefaultAsync(g => g.Id == teamMember.GroupId)).Name;
+            var memberUsername = (await _context.Users.FirstOrDefaultAsync(u => u.Id == memberUserId)).Username;
+
+            if (teamMember.InitiatorUserId == memberUserId)
```

Comment:
Try to avoid such big nested constructions

Reply:
Is it about if-else or getting name/username from db?